File: Day3/remove_duplicates_sorted_array.cpp
#include <iostream>
#include <vector>
using namespace std;

// Remove Duplicates from Sorted Array (Easy)
// Approach: Two pointers, in-place overwrite. Time O(n), Space O(1).
int removeDuplicates(vector<int>& nums) {
    if (nums.empty()) return 0;
    int write = 1;
    for (int read = 1; read < (int)nums.size(); ++read) {
        if (nums[read] != nums[write - 1]) {
            nums[write++] = nums[read];
        }
    }
    return write;
}

int main() {
    vector<int> nums{0,0,1,1,1,2,2,3,3,4};
    int k = removeDuplicates(nums);
    cout << "k = " << k << ", nums[:k] = [";
    for (int i = 0; i < k; ++i) {
        if (i) cout << ", ";
        cout << nums[i];
    }
    cout << "]\n";

    // basic checks
    vector<int> a; int k2 = removeDuplicates(a); if (k2 != 0) return 1;
    vector<int> b{1}; int k3 = removeDuplicates(b); if (!(k3==1 && b[0]==1)) return 1;
    vector<int> c{1,1,1}; int k4 = removeDuplicates(c); if (!(k4==1 && c[0]==1)) return 1;
    return 0;
}

File: Day3/remove_duplicates_sorted_array.py
"""
Remove Duplicates from Sorted Array (Easy)

Approach: Two pointers; overwrite in-place when encountering a new unique value.
Time: O(n), Space: O(1).
"""
from typing import List


def remove_duplicates(nums: List[int]) -> int:
    """Remove duplicates in-place, return length k of unique prefix.

    Modifies nums so that nums[:k] are the unique elements in order.
    """
    if not nums:
        return 0
    write = 1
    for read in range(1, len(nums)):
        if nums[read] != nums[write - 1]:
            nums[write] = nums[read]
            write += 1
    return write


if __name__ == "__main__":  # simple smoke tests
    arr = [0, 0, 1, 1, 1, 2, 2, 3, 3, 4]
    k = remove_duplicates(arr)
    assert k == 5 and arr[:k] == [0, 1, 2, 3, 4]

    arr2 = []
    k2 = remove_duplicates(arr2)
    assert k2 == 0 and arr2 == []

    arr3 = [1]
    k3 = remove_duplicates(arr3)
    assert k3 == 1 and arr3[:k3] == [1]

    arr4 = [1, 1, 1]
    k4 = remove_duplicates(arr4)
    assert k4 == 1 and arr4[:k4] == [1]

    print("remove_duplicates_sorted_array.py: all tests passed")
